test/gtest/ucm/malloc_hook.cc

```diff
@@ -1087,62 +1087,169 @@ class malloc_hook_dlopen : public malloc_hook {
     static std::string get_lib_path_do_mmap() {
         return get_lib_dir() + "/libdlopen_test_do_mmap.so";
     }
-};
 
-/* test for mmap events are fired from non-direct load modules
- * we are trying to load lib1, from lib1 load lib2, and
- * fire mmap event from lib2 */
-UCS_TEST_F(malloc_hook_dlopen, indirect_dlopen) {
-    typedef void (fire_mmap_f)(void);
-    typedef void* (load_lib_f)(const char *path);
+    static std::string get_lib_path_do_load_rpath() {
+        return get_lib_dir() + "/libdlopen_test_do_load_rpath.so";
+    }
 
-    const char *load_lib       = "load_lib";
-    const char *fire_mmap      = "fire_mmap";
+    static std::string get_lib_path_do_load_sub_rpath() {
+        return "libdlopen_test_rpath.so"; // library should be located using rpath
+    }
 
-    void *lib;
-    void *lib2;
-    load_lib_f *load;
-    fire_mmap_f *fire;
-    ucs_status_t status;
-    mmap_event<malloc_hook> event(this);
+    /* test for mmap events are fired from non-direct load modules
+     * we are trying to load lib1, from lib1 load lib2, and
+     * fire mmap event from lib2 */
+    template <void* (C)(const char*, int)>
```

Comment:
do we really need to use template? why not pass dlopen function as argument?

Reply:
in most of the cases templates simplify code (don't need to store additional parameters, check for NULL pointers), arguments for loader type can be defined in compile time.

but argument will work too, re-implemented